org.metaborg.spt.core/src/main/java/org/metaborg/spt/core/run/SpoofaxTestExpectationOutput.java

```diff
@@ -1,26 +1,51 @@
 package org.metaborg.spt.core.run;
 
 import org.metaborg.core.messages.IMessage;
-import org.metaborg.mbt.core.run.IFragmentResult;
-import org.metaborg.mbt.core.run.TestExpectationOutput;
-import org.metaborg.spoofax.core.unit.ISpoofaxAnalyzeUnit;
-import org.metaborg.spoofax.core.unit.ISpoofaxParseUnit;
 
-import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
-public class SpoofaxTestExpectationOutput extends TestExpectationOutput<ISpoofaxParseUnit, ISpoofaxAnalyzeUnit>
-    implements ISpoofaxTestExpectationOutput {
+public class SpoofaxTestExpectationOutput implements ISpoofaxTestExpectationOutput {
 
-    private final Iterable<ISpoofaxFragmentResult> fragmentResults;
+    private final boolean success;
+    private final List<IMessage> messages;
+    private final List<ISpoofaxFragmentResult> fragmentResults;
 
-    public SpoofaxTestExpectationOutput(boolean success, Iterable<IMessage> messages,
-        Iterable<ISpoofaxFragmentResult> fragmentResults) {
-        super(success, messages,
-            Lists.<IFragmentResult<ISpoofaxParseUnit, ISpoofaxAnalyzeUnit>>newLinkedList(fragmentResults));
-        this.fragmentResults = fragmentResults;
+    public SpoofaxTestExpectationOutput(boolean success, Collection<IMessage> messages, Collection<ISpoofaxFragmentResult> fragmentResults) {
+        this.success = success;
+        this.messages = copyFromCollection(messages);
+        this.fragmentResults = copyFromCollection(fragmentResults);
     }
 
-    @Override public Iterable<ISpoofaxFragmentResult> getFragmentResults() {
+    @Override public boolean isSuccessful() {
+        return success;
+    }
+
+    @Override public List<IMessage> getMessages() {
+        return messages;
+    }
+
+    @Override public List<ISpoofaxFragmentResult> getFragmentResults() {
         return fragmentResults;
     }
+
+    /**
+     * Copies the given collection.
+     *
+     * @param collection the collection to copy
+     * @param <T> the type of elements in the collection
+     * @return the copied collection
+     */
+    private static <T> List<T> copyFromCollection(Collection<T> collection) {
+        if (collection.isEmpty()) {
+            return Collections.emptyList();
+        } else if (collection.size() == 1) {
+            return Collections.singletonList(collection.iterator().next());
+        } else {
+            return Collections.unmodifiableList(new ArrayList<>(collection));
+        }
+    }
```

Comment:
Can you test if the given collection is already an unmodifiable list? That would save you an allocation/copy.

Reply:
No this would not work for two reasons. Firstly, there is no public unmodifiable list class for me to check with `instanceof`. I would have to check this with reflection or try to add something to the list and see it fail. Secondly, even if I test that it's an unmodifiable list, it can be a wrapper around a list that I don't own, so it might be modified somewhere else.

The goal of this method is to get a safe shallow copy of the list, so it actually doesn't matter that it's an unmodifiable list, since I own it and I never mutate it. I will remove the call to `Collections.unmodifiableList()`.